### DISCUSSION TITLE
Should a partial cartesian product using `HASH JOIN` use all cores?  How about if I split it up with `UNION ALL`?

### DISCUSSION THREAD

**Discussion by RobinL** [duckdb/duckdb#9710] [Q&A] [answered]
I'm the author of [Splink](http://github.com/moj-analytical-services/splink), software for data deduplication which uses DuckDB for its calculations.

A core calculation is a self-join of the input table.  The join is an `INNER JOIN` on low-cardinality column(s), resulting in a partial cartesian product (i.e. create all combinations of rows within each join key).  

I'm struggling to get this query to fully parallelise (use all cores) using duckdb 0.9.2, or any previous versions, and I'd be grateful for any pointers about:
- Whether this is expected behaviour
- Any suggestion that might cause better CPU usage

Here's a simplified example:

```sql
SELECT 

-- CPU-intensive string comparison operations.  In reality, there are many more of these
levenshtein(A.first_name, B.first_name) as lev,
jaro_winkler_similarity(A.first_name, B.first_name) as jw

from input_table as A
INNER JOIN input table as B
on A.year_of_birth = B.year_of_birth
```

This query does not seem to parallelise in DuckDB. 

<details><summary>Expand for a reproducible, runnable example using Python</summary>

```python
import duckdb

con = duckdb.connect(database=":memory:")

n = 1000


sql = f"""
CREATE TABLE test_table AS
SELECT
  range as unique_id,
  random() AS num,
  substr(characters, floor(random()*52)::integer + 1, 16) AS random_string_16,
  substr(join_key, floor(random()*4)::integer + 1, 1) AS join_key,
FROM
  range(0, {n:.0f}, 1),
  (SELECT 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz' AS characters) AS c,
  (SELECT 'ABCD' AS join_key) AS j
"""
con.execute(sql).commit()


query = """
-- Self-join table, creating all pairs of rows within each join_key group
-- Since join_key is low-cardinality this creates large number of rows
SELECT
  -- CPU intesive operations on the strings
  jaro_winkler_similarity(A.random_string_16, B.random_string_16) as jw_16,
  damerau_levenshtein(A.random_string_16, B.random_string_16) as dl_16,
  jaro_similarity(A.random_string_16, B.random_string_16) as j_16,
  jaccard(A.random_string_16, B.random_string_16) as jacc_16
FROM test_table as A
INNER JOIN test_table as B
on A.join_key = B.join_key
WHERE A.unique_id < B.unique_id
"""
con.execute("SET explain_output='all';").commit()
explain_query = f"EXPLAIN {query}"
execution_plan = con.execute(explain_query).fetchall()
for plan_type in execution_plan:
    for plan in plan_type:
        print(plan)


con.execute(f"create table A as {query}").commit()
print(con.execute(f"select count(*) from A").df())
con.execute(f"drop table A").commit()
```

</details>

One idea I had was to split the calculation into multiple `UNION ALLs`, like so:

```sql
SELECT 
levenshtein(A.first_name, B.first_name) as lev,
jaro_winkler_similarity(A.first_name, B.first_name) as jw
from input_table as A
INNER JOIN input table as B
on A.year_of_birth = B.year_of_birth
WHERE A.unique_id % 2 = 0

UNION ALL

SELECT 
levenshtein(A.first_name, B.first_name) as lev,
jaro_winkler_similarity(A.first_name, B.first_name) as jw
from input_table as A
INNER JOIN input table as B
on A.year_of_birth = B.year_of_birth
WHERE A.unique_id % 2 = 1
```

Unfortunately this doesn't seem to help.  

What follows is a more extended example that demonstrates these queries don't seem to parallelise, even with the `UNION ALL` strategy

<details><summary>Extended example</summary>

```python

import duckdb

import time

con = duckdb.connect(database=":memory:")



n = 1000
sql = f"""
CREATE TABLE test_table AS
SELECT
  random() AS num,
  substr(characters, floor(random()*62)::integer + 1, 4) AS random_string_4,
  substr(characters, floor(random()*62)::integer + 1, 8) AS random_string_8,
  substr(characters, floor(random()*62)::integer + 1, 16) AS random_string_16,
  substr(join_key, floor(random()*4)::integer + 1, 1) AS join_key,
FROM
  range(0, {n:.0f}, 1),
  (SELECT 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789' AS characters) AS c,
  (SELECT 'ABCD' AS join_key) AS j
"""
con.execute(sql).commit()

con.execute("SELECT * from test_table limit 2").df()


times = []

for i in [0, 1, 2, 4, 8, 16, 32]:
    if i != 0:
        con.execute(
            f"""
        SET threads TO {i};
        SET enable_progress_bar=true;
        SET external_threads TO {i};
        SET worker_threads to {i};
        """
        ).commit()

    start_time = time.time()

    query_template = """
    SELECT
      jaro_winkler_similarity(A.random_string_16, B.random_string_16) as jw_16,
      damerau_levenshtein(A.random_string_16, B.random_string_16) as dl_16,
      jaro_similarity(A.random_string_16, B.random_string_16) as j_16,
      jaccard(A.random_string_16, B.random_string_16) as jacc_16,
    FROM test_table as A
    INNER JOIN test_table as B
    on A.join_key = B.join_key
    where floor(A.num*{num_partitions}) = {partition}
    """

    num_partitions = 8

    queries = [query_template.format(partition=i, num_partitions=16) for i in range(16)]
    query = " UNION ALL ".join(queries)
    print(queries[0])
    con.execute(f"create table A as {query}").commit()
    con.execute(f"drop table A").commit()

    end_time = time.time()

    times.append(end_time - start_time)
    print(times)

print(f"Average times: {times}")
```

</details>

Output of the above:

```
Times: [4.367391109466553, 4.315721035003662, 4.373572111129761, 4.372523069381714, 4.274524927139282, 4.297921895980835, 4.27379035949707]
```

(i.e. the time taken is the same irrespective of the `thread` PRAGMA.)

Super grateful for any advice

Finally, just for comparison, I note that I have been able to achieve full parallelisation on a more simple query, so I don't think there's anything wrong with my setup

<details><summary>simple example of full parallelisation</summary>

```python
import duckdb
import pandas as pd
import time



# Create a DuckDB connection with threads set to use all available cores
con = duckdb.connect(database=":memory:", read_only=False)

n = 5e7
con.execute(
    f"CREATE TABLE test_table as SELECT random() as num FROM range(0, {n:.0f}, 1)"
)


times = []

for i in [0, 1, 2, 4, 8, 16, 32]:
    if i != 0:
      con.execute(
          f"""
      SET threads TO {i};
      SET enable_progress_bar=true;
      SET external_threads TO {i};
      SET worker_threads to {i};
      """
      ).commit()

    start_time = time.time()

    # Create a large table with generated series and random numbers

    # This query should def parallelise well?
    query = """
    SELECT num,
           sin(num) as sin_val,
           log(num + 1) as log_val,
           exp(num) as exp_val,
           sqrt(num) as sqrt_val
    FROM test_table
    """

    " UNION ALL ".join([query * 8])
    con.execute(f"create table A as {query}").commit()
    con.execute(f"drop table A").commit()


    end_time = time.time()

    times.append(end_time - start_time)
    print(times)

print(f"Average times: {times}")

# The above seems to work and uses 100% of cores
# Average times: [1.8005571365356445, 7.63769793510437, 3.9917047023773193, 2.2830631732940674, 1.8552842140197754, 1.8657467365264893, 1.7939019203186035]

```
</details>

[Here](https://github.com/duckdb/duckdb/issues/2653) is a related issue, which suggests that perhaps my queries should be parallelising but I'm not sure.  [Here](https://github.com/duckdb/duckdb/pull/3700) is a PR that may be relevant (Support Parallel Order-Preserving Result Set Materialization), but I don't understand the terminology well enough to know whether it's relevant

**Comment by Mytherin** [comment, marked as answer]
The regular query doesn't parallelize because the table is very small, and the data set size is increased by an exploding join. DuckDB parallelizes based on the base table sizes so if a table has a size of 1000 tuples it will only ever use a single thread regardless of how many tuples are generated in subsequent exploding joins.

The union variant not parallelizing is indeed related to the output order issue - we currently only execute `UNION` pipelines in parallel if order is not required in the final stage, adding an `ORDER BY` resolves the issue, e.g.:

```py
con.execute(f"create table A as {query} order by 1").commit()
```

**Reply by RobinL** [reply]
Thank you so much.  Can confirm the `order by 1` immediately resulted in parallelisation for the `UNION ALL` query.

I wondered if you could provide more info (maybe just a link to the code) on where a 'small table' is defined?  

Our 'real life' input tables are fairly small, but not tiny (up to about 1m records - but for for the non-`UNION ALL` variant, I've never seen full parallelisation in any of our workloads. 

Finally, other than the `UNION ALL` Do you know any 'tricks' that would trigger parallelisation of the exploding join query?

p.s this is hugely exciting to us!  duckdb is already outperforming heavily parallelised spark even with the single-threaded query.  I've just tested this against real data and for inference it gives me an immediate 4x speedup on a 6 core Macbook!

**Reply by Mytherin** [reply]
The unit of parallelism is a single row group (120K rows), so with 1M rows you should see up to 8 cores being utilized. 

> Finally, other than the UNION ALL Do you know any 'tricks' that would trigger parallelisation of the exploding join query?

You could materialize the result of the join in a table (or perhaps even a materialized CTE).

**Reply by RobinL** [reply]
Thank you so much - I will experiment with those options and see what works best for us

**Reply by zmbc** [reply]
> adding an ORDER BY resolves the issue

Am I missing something, or does ordering the table require a bunch more computation, that doesn't parallelize especially well? I expected that setting [`preserve_insertion_order`](https://duckdb.org/docs/sql/configuration.html) to False would get me the same benefit as the ORDER BY without the extra work, but it doesn't seem to.

**Reply by Mytherin** [reply]
This has to do with the way streaming query results work - which are triggered by `con.execute()` in Python. By default we stream the result to the client when `con.execute` is used as it forces us to execute before we know how the user wants to read the result - but this streaming causes the final result fetching/computation to be executed in a single thread in the current implementation.

Adding an `ORDER BY` forces the query to materialize the query result as part of the query, which is all done in parallel. The single-threaded result reading is then fast because the result has already been computed and stored as part of executing the `ORDER BY`.

Using `con.sql()` is the preferred method of using the Python client for this reason as it does not suffer from this problem.

We're also shipping parallel streaming query results with the next version of DuckDB which should address this problem as well and enable parallel computation of streaming query results, see https://github.com/duckdb/duckdb/pull/10245.

**Reply by RobinL** [reply]
Thank you - that is super helpful

**Reply by zmbc** [reply]
Amazing! @Mytherin I'm using DuckDB via Ibis (and therefore via SQLAlchemy) so it isn't easy for me to switch to `con.sql()`, but it sounds like this will all start working for me with the next version 🎉

**Reply by NickCrews** [reply]
Chiming in that I am also using ibis as an interface to duckdb, so any way to make optimizations work via that interface would be sweet!

Also noting that Ibis is [moving to using sqlglot instead of sqlalchemy](https://github.com/ibis-project/ibis/issues/7909) as their translation layer, if that makes some difference here.
